hathor/conf/get_settings.py

```diff
@@ -37,52 +37,82 @@ def HathorSettings() -> Settings:
-        return None
-
-    try:
-        module = importlib.import_module(config_file)
-    except ModuleNotFoundError:
-        default_file = 'hathor.conf.mainnet'
-        module = importlib.import_module(default_file)
+    XXX: This is only meaningful after HathorSettings() is used at least once.
+    """
+    return _config_file
 
-    return module
 
+def get_settings_filepath() -> Optional[str]:
+    """ Returns the path of the YAML settings loaded, None if no YAML file was loaded.
 
-def get_settings_filepath() -> str:
-    global _settings_filepath
+    XXX: This is only meaningful after HathorSettings() is used at least once.
+    """
+    return _settings_filepath
 
-    new_settings_filepath = os.environ.get('HATHOR_CONFIG_YAML', conf.MAINNET_SETTINGS_FILEPATH)
 
-    if _settings_filepath is not None and _settings_filepath != new_settings_filepath:
-        raise Exception('loading config twice with a different file')
+def _get_module_settings_filepath(env_name: str) -> Optional[str]:
+    return os.environ.get(env_name)
 
-    _settings_filepath = new_settings_filepath
 
-    return new_settings_filepath
+def _load_module_settings_with_cache(module_path: str) -> Settings:
+    global _config_file
+    global _settings_filepath
+    global _settings_singleton
+    if _settings_singleton is not None:
+        if _config_file != module_path:
+            raise Exception('loading config twice with a different file')
+        if _settings_filepath is not None:
+            raise Exception('loading config twice with a different file type')
+        return _settings_singleton
+    settings = _load_module_settings(module_path)
+    _config_file = module_path
+    _settings_singleton = settings
+    return settings
+
+
+def _load_module_settings(module_path: str) -> Settings:
+    log = logger.new()
+    log.warn(
+        "Setting a config module via the 'HATHOR_CONFIG_FILE' env var will be deprecated soon. "
+        "Use the '--config-yaml' CLI option or the 'HATHOR_CONFIG_YAML' env var to set a yaml filepath instead."
+    )
+    settings_module = importlib.import_module(module_path)
+    settings = getattr(settings_module, 'SETTINGS')
+    assert isinstance(settings, Settings)
+    return settings
+
+
+def _get_yaml_settings_filepath(env_name: str) -> str:
+    return os.environ.get(env_name, conf.MAINNET_SETTINGS_FILEPATH)
+
+
+def _load_yaml_settings_with_cache(filepath: str) -> Settings:
+    global _config_file
+    global _settings_filepath
+    global _settings_singleton
+    if _settings_singleton is not None:
+        if _settings_filepath != filepath:
+            raise Exception('loading config twice with a different file')
+        if _config_file is not None:
+            raise Exception('loading config twice with a different file type')
+        return _settings_singleton
+    settings = _load_yaml_settings(filepath)
+    _settings_filepath = filepath
+    _settings_singleton = settings
+    return settings
```

Comment:
I found it a bit weird that we needed 3 global vars, and also that `_load_yaml_settings_with_cache()` is very similar to `_load_module_settings_with_cache()`.

If we implement a structure like `SettingsMetadata` below, we can have a single global var and eliminate duplicate code:

```python
class SettingsMetadata(NamedTuple):
    source: str
    is_yaml: bool
    settings: Settings

_settings_metadata_singleton: Optional[SettingsMetadata] = None

def _load_settings_with_cache(source: str, *, is_yaml: bool) -> Settings:
    global _settings_metadata_singleton

    if _settings_metadata_singleton is not None:
        if _settings_metadata_singleton.is_yaml != is_yaml:
            raise Exception('loading config twice with a different file type')
        if _settings_metadata_singleton.source != source:
            raise Exception('loading config twice with a different file')

        return _settings_metadata_singleton.settings

    settings_loader = _load_yaml_settings if is_yaml else _load_module_settings
    _settings_metadata_singleton = SettingsMetadata(
        source=source,
        is_yaml=is_yaml,
        settings=settings_loader(source)
    )

    return _settings_metadata_singleton.settings
```

Or something like this. What do you think?

Reply:
That seems good, I'll do that.

Reply:
Great, thanks!